### ISSUE TITLE
Problem with ZEXT SEXT and TRUNC

### ISSUE THREAD

**Issue by wutianba** [NASA-SW-VnV/ikos#10]
```c++
extern int random();
unsigned char cnt;
void main() {
  int A[8];
  cnt = random();
  if (cnt > 0 && cnt < 8) { 
    A[cnt] = 0;
  }
}
```
In this case, we can't find relationships between pre-conversion and post-conversion because the implicit conversion produces temporary variables.
I used to convert `a = ZEXT(b)` (`SEXT` or `TRUNC`) to `a = b` or add `a equals b` relation to the relational abstraction when it is sound.
Is there any other better way or we should modify current transfer functions?

**Comment by arthaud** [comment]
Hi @wutianba,

This is a problem that I have seen before. I believe there are actually 2 separate issues.

First, there is a problem with zext/sext/trunc. As you said, `cnt > 0` turns into `tmp = zext(cnt); tmp > 0;` and IKOS cannot infer that `cnt > 0` with the interval domain. I have fixed this by writing a pass  called [SimplifyUpcastComparisonPass](https://github.com/NASA-SW-VnV/ikos/blob/master/ar/include/ikos/ar/pass/simplify_upcast_comparison.hpp) that adds an extra comparison on the original variable: `cnt > 0`. This pass is trivially sound and fixes the issue.

For instance, if I use a local variable instead of a global variable:
```c
extern int random();

int main() {
  int A[8];
  unsigned char cnt = random();
  if (cnt > 0 && cnt < 8) { 
    A[cnt] = 0;
  }
  return 0;
}
```

The Abstract Representation is:
```
define si32 @main() {
#1 !entry successors={#2, #3} {
  [8 x si32]* $1 = allocate [8 x si32], 1, align 16
  si32 %2 = call @random()
  ui32 %3 = bitcast %2
  ui8 %4 = utrunc %3
  ui32 %5 = zext %4
  si32 %6 = bitcast %5
}
#2 predecessors={#1} successors={#4, #5} {
  %6 sigt 0
  %4 uigt 0
  ui32 %7 = zext %4
  si32 %8 = bitcast %7
}
#3 predecessors={#1} successors={#6} {
  %6 sile 0
  %4 uile 0
}
#4 predecessors={#2} successors={#6} {
  %8 silt 8
  %4 uilt 8
  ui64 %9 = zext %4
  si32* %10 = ptrshift $1, 32 * 0, 4 * %9
  store %10, 0, align 4
}
#5 predecessors={#2} successors={#6} {
  %8 sige 8
  %4 uige 8
}
#6 !exit predecessors={#3, #5, #4} {
  return 0
}
}
```

Notice that the comparisons are duplicated, and IKOS is able to prove that this code is safe.

The second problem is due to the indirection produced by the global variable. In LLVM, a global variable is represented by an address, every access to it is translated into a memory load or store. So with your code, we actually get `tmp1 = *g; if (tmp1 > 0 && tmp1 < 8) { tmp2 = *g; A[tmp2] = 0; }`. IKOS infers properties on `tmp1` but cannot relate it to `*g`. I don't think we can fix this without using a relational abstract domain.

PS: The pass was actually implemented by @zadlg

**Comment by arthaud** [comment]
In theory, the false positive should go away when using a relational abstract domains, such as dbm or polyhedra.
In practice, `ikos test.c -d=dbm` and `ikos test.c -d=apron-ppl-polyhedra` still produce the warning.
I need to investigate why, when I will have time..

**Comment by arthaud** [comment]
This is now fixed in 224fcee4b4c369666c7f50d84097247294d70b64.

You will have to use a relational abstract domain.
For instance, you can use `dbm` or `apron-ppl-polyhedra`.
For bigger code, I would recommend `var-pack-dbm`. It scales pretty well and has a good precision.
